Client assertion backend (#82)
bluesky-social/proposals: 0010-client-assertion-backend/README.md

# Proposal: Client assertion backend for browser-based applications

> This proposal written by Devin Ivy and edited Matthieu Sieben, on behalf of Bluesky. First posted in June 2025.

> In its current form, ATProto's OAuth flavor requires app developers to make a choice: use a “confidential” or “public” client. This choice will impact the security properties of the OAuth session like its lifetime, the ability to use silent-sign-on, etc.

Reviewer: Aligning with terminology from our spec.
```suggestion
In its current form, ATProto's OAuth flavor requires app developers to make a choice: use a “confidential” or “public” client. This choice will impact the security properties of the OAuth session like its lifetime, the ability to use silent sign-in, etc.
```

Single page apps (web browser) and native apps, are applications that run and hold credentials on user devices. They typically don’t have a server component (besides a trivial one for serving static assets), making them, defacto, “public”. In order to become “confidential”, clients MUST be able to securely maintain, and use, private keys. This means that they MUST rely on a more complex server component (back-end).

Native and browser-based apps often achieve this through a [Token Mediating Backend](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-browser-based-apps#section-6.2.1) (TMB) architecture. But implementing such a server can be quite complex, as the server will have to deal with session management, token storage, etc. This is especially true when combined with DPoP.

> The purpose of this proposal is to outline a low-footprint method of making confidential browser-based clients. It does so by extending the [JWT Profile for OAuth 2.0 Client Authentication and Authorization Grants](https://datatracker.ietf.org/doc/html/rfc7523) – when used in combination with [Demonstrating Proof of Possession](https://datatracker.ietf.org/doc/html/rfc9449).

Reviewer: ```suggestion
The purpose of this proposal is to outline a low-footprint method of making confidential browser-based clients. It does so by extending the [JWT Profile for OAuth 2.0 Client Authentication and Authorization Grants](https://datatracker.ietf.org/doc/html/rfc7523) – when used in combination with [Demonstrating Proof of Possession](https://datatracker.ietf.org/doc/html/rfc9449), which is already a part of the ATProto OAuth profile.
```

  "client_assertion": "eyJhbGciOiJFUzI1NiJ9.eyJpc3MiOiJodHRwczovL215LmNsaWVudC5jb20vb2F1dGgtY2xpZW50LW1ldGFkYXRhLmpzb24iLCJzdWIiOiJodHRwczovL215LmNsaWVudC5jb20vb2F1dGgtY2xpZW50LW1ldGFkYXRhLmpzb24iLCJhdWQiOiJodHRwczovL2Jza3kuc29jaWFsIiwianRpIjoiZTBxOHp1YTJxdW4iLCJpYXQiOjE3NDkwMjE2MDksImNuZiI6eyJqa3QiOiIwN1FERUZRcVduV0ZCUmxCOFprN09idzFJMktabFhBb0ZqeGJicjd0Vm9BIn19.vgYiLDqYXeCGHxJdl_eYGpqeWfU3EPQh8Dv7BMBnh9OBGET0Sr2Wj_y7ViJsxMC_X2YeQZICIehQbHYG9wOMPw"
}
```

## Conclusion

By having client key material, sessions may be revoked en masse by rotating the keys advertised in the client’s metadata document.

Individual sessions may be discontinued by the client assertion backend refusing to issue new assertions for a given DPoP key.

Max session lifetimes or other arbitrary session lifetime rules can be enforced by the client assertion backend.

By bounding the client authentication to the session DPoP key it greatly constrains the impact of having an open endpoint for providing client assertions (though the backend may also require other forms of authentication). You could not mint an assertion in one session then use it in another.

> This essentially puts a client backend “in the middle” of credential issuance from the token endpoint, without the token request actually having to pass through the backend service. This is in contrast to the token-mediating backend pattern, which acts more like a proxy with its own auth lifecycle.

Reviewer: ```suggestion
This essentially gives a client backend veto power over token refresh, without the token request actually having to pass through the backend service. This is in contrast to the token-mediating backend pattern, which acts more like a proxy with its own auth lifecycle.
```
